### PR TITLE
ENH: Improve performance for arrow dtypes in monotonic join

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

```
idx = Index(list(range(1, 1_000_000)), dtype="int64[pyarrow]")
idx2 = Index(list(range(100_000, 1_100_000)), dtype="int64[pyarrow]")
idx.union(idx2)

# main
# %timeit idx.union(idx2)
# 327 ms ± 72.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

# pr
# %timeit idx.union(idx2)
# 2.79 ms ± 27.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

### CODE REVIEW

**Comment by jbrockmendel** [comment]
no objection here, but eventually we ought to find a way to do this dispatch without special-casing inside the index code (i.e. implement something at the EA level)

hows is perf affected on multi-chunk pyarrow objs?

**Comment by phofl** [comment]
Arrays have 2 million entries, initial performance 380ms, on this pr

```
%timeit idx.union(idx2)
11.5 ms ± 477 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

I totally agree with your point regarding EA interface, this is a short term solution for 2.0

**Comment by phofl** [comment]
@jbrockmendel ok to merge?

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -4948,6 +4955,10 @@ def _from_join_target(self, result: np.ndarray) -> ArrayLike:
         """
         if isinstance(self.values, BaseMaskedArray):
             return type(self.values)(result, np.zeros(result.shape, dtype=np.bool_))
+        elif isinstance(self.values, ArrowExtensionArray):
+            import pyarrow as pa
+
+            return type(self.values)(pa.array(result))
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:4961]: _from_sequence?

**Review by jbrockmendel — APPROVED** [review]
LGTM

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (phofl) [reply pandas/core/indexes/base.py]: good point, changed

**Pull request merged**